# # SWAM - Smartly weighted averaging across multiple tissues

## Overview

SWAM is an gene expression imputation method which combines information from multiple sources to boost accuracy of imputed gene expression levels 

## Getting Started

### Prerequisites

To use SWAM, you need to have the following tools installed in your system.
* `perl` (version 5 or later recommended)
* `R` (version 3.4 or later recommended)
* `htslib` : The binary `tabix` should be include in your `$PATH`. Type `tabix` in your command line to check.
* `prediXcan` : The software `prediXcan` can be downlaoded from https://github.com/hakyimlab/PrediXcan
* `sqlite` : (version 3 or later recommended)


### Cloning the repository

Next, clone the repository using the following command:

```
git clone https://github.com/aeyliu/SWAM.git
```

### Preparing Input Files

* genotypes : the genotype file should be dosages with the following columns - chromosome, RSID, position, reference allele, alternate allele, minor allele frequency, dosages
* samples : a text file with sample IDs should be included corresponding to the columns of the genotype file
* single-tissue database files : these .db files are compatible with the format generated by prediXcan (for example, GTEx derived tissues can be obtained from http://predictdb.org/)
* expression file for target tissue : measured expression file with following columns - gene ensemble ID, expression values (for each sample)


### Running SWAM (example)



## Frequently Asked Questions (FAQs)


## Citing SWAM
